File: sdk-ios/Sources/WailoSDK/WailoAddress.swift
import Foundation
import Network

/// A desktop address known to be diallable: a host that is legal inside a URL authority, plus the port
/// the text named (`nil` when it named none, leaving the choice to the caller).
///
/// Every address here arrives as free text — the on-device panel's field, a `-WailoHost` launch
/// argument, a literal passed to `Wailo.start` — and `URL(string:)` returns nil for text it cannot
/// parse. Typing an IP together with its port into the address field yields `ws://10.0.0.2:8080:8899/`,
/// which is one of those, so the transport's force-unwrap turned a typo into a crash of the *host app*;
/// and since the text was persisted before it was ever dialled, the crash then repeated on every launch.
/// Parsing is therefore the one gate in front of `UserDefaults` and the transport: text that names no
/// diallable address is refused, never stored.
///
/// Forgiving about shape, strict about legality. `host:port`, a bare IPv6 literal, and a pasted
/// `ws://host:port/` all resolve, because each is a plausible thing to type and none is ambiguous.
/// Anything Foundation would reject is refused up front rather than dialled and left to look like "the
/// desktop isn't running".
public struct WailoAddress: Equatable, Sendable, CustomStringConvertible {

    public static let portRange = 1...65_535

    /// Ready to drop into a URL authority — an IPv6 literal keeps the brackets it needs there.
    public let host: String

    /// The port the text named, or `nil` for "whatever the caller resolves to": a discovered service's
    /// port, a separately configured one, or `Wailo.defaultPort`.
    public let port: Int?

    /// Canonical `host` or `host:port`, which re-parses to the same address.
    public var description: String { port.map { "\(host):\($0)" } ?? host }

    public init?(_ text: String) {
        guard let (host, portText) = Self.split(Self.authority(of: text)) else { return nil }

        var port: Int?
        if let portText {
            guard let value = Int(portText), Self.portRange.contains(value) else { return nil }
            port = value
        }
        guard Self.isDiallable(host: host, port: port) else { return nil }

        self.host = host
        self.port = port
    }

    /// The `ws://` URL for a host and port. Optional on purpose, and the only place in the SDK that
    /// turns an address into a URL: `URL(string:)` failing here is the crash this type exists to
    /// prevent, so no caller gets to force it.
    static func webSocketURL(host: String, port: Int) -> URL? {
        URL(string: "ws://\(host):\(port)/")
    }

    // MARK: - parsing

    /// Reduces a pasted address to its `host[:port]` authority. Addresses get copied from somewhere —
    /// Studio prints its own as a URL — so a scheme and a trailing slash are decoration, not intent.
    private static func authority(of text: String) -> String {
        var authority = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if let scheme = authority.range(of: "://") {
            authority = String(authority[scheme.upperBound...])
        }
        while authority.hasSuffix("/") {
            authority.removeLast()
        }
        return authority
    }

    private static func split(_ authority: String) -> (host: String, port: String?)? {
        guard !authority.isEmpty else { return nil }

        if authority.hasPrefix("[") {
            guard let close = authority.firstIndex(of: "]") else { return nil }
            let host = String(authority[...close])
            let rest = authority[authority.index(after: close)...]
            if rest.isEmpty { return (host, nil) }
            guard rest.hasPrefix(":") else { return nil }
            return (host, String(rest.dropFirst()))
        }

        let parts = authority.split(separator: ":", omittingEmptySubsequences: false)
        switch parts.count {
        case 1:
            return (authority, nil)
        case 2:
            return (String(parts[0]), String(parts[1]))
        default:
            // Several colons and no brackets reads as an IPv6 literal typed bare. Bracketing is what
            // makes it a legal authority, and a port could not have been expressed without them —
            // `isDiallable` is what rejects the other reading of the same text, a doubled port.
            return ("[\(authority)]", nil)
        }
    }

    private static let illegalInHost = CharacterSet.whitespacesAndNewlines
        .union(CharacterSet(charactersIn: "/?#@[]\\"))

    private static func isDiallable(host: String, port: Int?) -> Bool {
        if host.hasPrefix("[") {
            // Foundation takes a bracketed authority on trust without parsing what is inside, so a
            // `10.0.0.2:8080:9000` double-port typo would otherwise pass as an IPv6 literal.
            guard host.hasSuffix("]"),
                  IPv6Address(String(host.dropFirst().dropLast())) != nil else { return false }
        } else if host.isEmpty || host.rangeOfCharacter(from: illegalInHost) != nil {
            return false
        }
        // Foundation decides what it will parse, so an address is valid exactly when the URL it forms
        // exists. This is the check whose absence crashed the app.
        return webSocketURL(host: host, port: port ?? Wailo.defaultPort) != nil
    }
}

File: sdk-android/src/main/kotlin/com/venbiasa/wailo/sdk/android/WailoAddress.kt
package com.venbiasa.wailo.sdk.android

import java.net.URI

/**
 * A desktop address known to be diallable: a host legal inside a URL authority, plus the port the text
 * named (null when it named none, leaving the choice to the caller).
 *
 * Every address arrives as free text — the on-device panel's field, a literal passed to `Wailo.start`,
 * an intent extra — so this is the one gate in front of persistence and the transport. Typing an IP
 * together with its port into a field labelled "address" is the input that has to be caught: on iOS it
 * produced `ws://10.0.0.2:8080:8899/` and crashed the host app, and because the text was stored before
 * it was ever dialled, the crash then repeated on every launch (ADR-0035). Text that names no diallable
 * address is refused here rather than saved and left to look like "the desktop isn't running".
 *
 * Forgiving about shape, strict about legality. `host:port`, a bare IPv6 literal, and a pasted
 * `ws://host:port/` all resolve, because each is a plausible thing to type and none is ambiguous.
 */
class WailoAddress private constructor(
    /** Ready to drop into a URL authority — an IPv6 literal keeps the brackets it needs there. */
    val host: String,
    /** The port the text named, or null for "whatever the caller resolves to". */
    val port: Int?,
) {

    /** Canonical `host` or `host:port`, which re-parses to the same address. */
    override fun toString(): String = port?.let { "$host:$it" } ?: host

    override fun equals(other: Any?): Boolean =
        other is WailoAddress && other.host == host && other.port == port

    override fun hashCode(): Int = host.hashCode() * 31 + (port ?: 0)

    companion object {
        val PORT_RANGE = 1..65_535

        fun parse(text: String): WailoAddress? {
            val (host, portText) = split(authority(text)) ?: return null
            val port = portText?.let { it.toIntOrNull()?.takeIf(PORT_RANGE::contains) ?: return null }
            if (!isDiallable(host)) return null
            return WailoAddress(host, port)
        }

        /**
         * Reduces a pasted address to its `host[:port]` authority. Addresses get copied from somewhere —
         * Studio prints its own as a URL — so a scheme and a trailing slash are decoration, not intent.
         */
        private fun authority(text: String): String {
            var authority = text.trim()
            val scheme = authority.indexOf("://")
            if (scheme >= 0) authority = authority.substring(scheme + 3)
            return authority.trimEnd('/')
        }

        private fun split(authority: String): Pair<String, String?>? {
            if (authority.isEmpty()) return null

            if (authority.startsWith("[")) {
                val close = authority.indexOf(']')
                if (close < 0) return null
                val host = authority.substring(0, close + 1)
                val rest = authority.substring(close + 1)
                if (rest.isEmpty()) return host to null
                if (!rest.startsWith(":")) return null
                return host to rest.substring(1)
            }

            val parts = authority.split(":")
            return when (parts.size) {
                1 -> authority to null
                2 -> parts[0] to parts[1]
                // Several colons and no brackets reads as an IPv6 literal typed bare. Bracketing is what
                // makes it a legal authority, and a port could not have been expressed without them —
                // the literal check below is what rejects the other reading of the same text, a doubled
                // port such as `10.0.0.2:8080:9000`.
                else -> "[$authority]" to null
            }
        }

        private const val ILLEGAL_IN_HOST = "/?#@[]\\"

        private fun isDiallable(host: String): Boolean {
            if (host.startsWith("[")) return host.endsWith("]") && isIpv6Literal(host)
            return host.isNotEmpty() && host.none { it.isWhitespace() || it in ILLEGAL_IN_HOST }
        }

        /**
         * `URI` only hands back a host for a bracketed authority when what is inside is a real IPv6
         * literal, so asking it is the check — rolling our own parser for a form this exact would be
         * a second, worse implementation of RFC 2732.
         */
        private fun isIpv6Literal(bracketed: String): Boolean =
            runCatching { URI("ws://$bracketed:1/").host }.getOrNull() == bracketed
    }
}
